src/subtype.c

```diff
@@ -1165,6 +1165,13 @@ JL_DLLEXPORT int jl_types_equal(jl_value_t *a, jl_value_t *b)
     return jl_subtype(a, b) && jl_subtype(b, a);
 }
 
+JL_DLLEXPORT int jl_is_not_broken_subtype(jl_value_t *a, jl_value_t *b)
+{
+    // TODO: the final commented out check here isn't correct; it should be closer to the
+    // `issingletype` check used by `isnotbrokensubtype` in `base/compiler/typeutils.jl`
+    return !jl_is_kind(b) || !jl_is_type_type(a); // || jl_is_datatype_singleton((jl_datatype_t*)jl_tparam0(a));
+}
+
```

Comment:
I've added this check before the problematic subtyping check in `emit_isa` and [it allows the tests in the previously provided gist](https://gist.github.com/jrevels/4e56e78d91d53ca29af73b5b36d506d0) to pass, including the final heisenbuggy one. It's emulating the `isbrokensubtype` predicate we're using in the `isa` tfunc for this PR. As you can see by the TODO, this predicate is probably overly strict. 

BTW, thanks to @maleadt, @DokFaust, @pablosanjose for looking into this. Much appreciated! 

Would also be nice to get some review from core codegen contributors. If tests pass here, then I'm going to merge right away unless there are any objections.